autoload/ale/engine.vim

```diff
@@ -72,22 +72,34 @@ function! s:StopPreviousJobs(buffer, linter) abort
     let g:ale_buffer_info[a:buffer].job_list = l:new_job_list
 endfunction
 
-function! s:GatherOutput(job, data) abort
-    let l:job_id = s:GetJobID(a:job)
+function! s:GatherOutputVim(channel, data) abort
+    let l:job_id = s:GetJobID(ch_getjob(a:channel))
 
     if !has_key(s:job_info_map, l:job_id)
         return
     endif
 
-    call extend(s:job_info_map[l:job_id].output, a:data)
-endfunction
-
-function! s:GatherOutputVim(channel, data) abort
-    call s:GatherOutput(ch_getjob(a:channel), [a:data])
+    call add(s:job_info_map[l:job_id].output, a:data)
 endfunction
 
 function! s:GatherOutputNeoVim(job, data, event) abort
-    call s:GatherOutput(a:job, a:data)
+    let l:job_id = s:GetJobID(a:job)
+
+    if !has_key(s:job_info_map, l:job_id)
+        return
+    endif
+
+    " Join the lines passed to ale, because Neovim splits them up.
+    " a:data is a list of strings, where every item is a new line, except the
+    " first one, which is the continuation of the last item passed last time.
+    if empty(s:job_info_map[l:job_id].output)
+        let s:job_info_map[l:job_id].output = a:data
+    else
+        let s:job_info_map[l:job_id].output =
+        \   s:job_info_map[l:job_id].output[:-2]
+        \   + [s:job_info_map[l:job_id].output[-1] . get(a:data, 0, '')]
+        \   + a:data[1:]
```

Comment:
I think you should write something like this instead.

```vim
" Extend the previous line, which can be continued.
let s:job_info_map[l:job_id].output[-1] .= get(a:data, 0, '')

" Add the new lines.
call extend(s:job_info_map[l:job_id].output, a:data[1:])
```

That will probably allow for better allocation for the List where the List can be extended, instead of creating new List instances each every time a line is added. The comment above should help explain what is going on better.